src/backend/distributed/commands/database.c

```diff
@@ -242,3 +274,421 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
+	if (!HeapTupleIsValid(tuple))
+	{
+		return NULL;
+	}
+
+	Form_pg_tablespace tablespaceForm = (Form_pg_tablespace) GETSTRUCT(tuple);
+	char *tablespaceName = pstrdup(NameStr(tablespaceForm->spcname));
+
+	ReleaseSysCache(tuple);
+
+	return tablespaceName;
+}
+
+
+/*
+ * GetDatabaseCollation gets oid of a database and returns all the collation related information
+ * We need this method since collation related info in Form_pg_database is not accessible.
+ */
+static DatabaseCollationInfo
+GetDatabaseCollation(Oid dbOid)
+{
+	DatabaseCollationInfo info;
+	memset(&info, 0, sizeof(DatabaseCollationInfo));
+
+	Relation rel = table_open(DatabaseRelationId, AccessShareLock);
+	HeapTuple tup = get_catalog_object_by_oid(rel, Anum_pg_database_oid, dbOid);
+	if (!HeapTupleIsValid(tup))
+	{
+		elog(ERROR, "cache lookup failed for database %u", dbOid);
+	}
+
+	bool isNull = false;
+
+	TupleDesc tupdesc = RelationGetDescr(rel);
+
+	Datum collationDatum = heap_getattr(tup, Anum_pg_database_datcollate, tupdesc,
+										&isNull);
+	info.datcollate = TextDatumGetCString(collationDatum);
+
+	Datum ctypeDatum = heap_getattr(tup, Anum_pg_database_datctype, tupdesc, &isNull);
+	info.datctype = TextDatumGetCString(ctypeDatum);
+
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+	Datum icuLocaleDatum = heap_getattr(tup, Anum_pg_database_daticulocale, tupdesc,
+										&isNull);
+	if (!isNull)
+	{
+		info.daticulocale = TextDatumGetCString(icuLocaleDatum);
+	}
+
+	Datum collverDatum = heap_getattr(tup, Anum_pg_database_datcollversion, tupdesc,
+									  &isNull);
+	if (!isNull)
+	{
+		info.datcollversion = TextDatumGetCString(collverDatum);
+	}
+#endif
+
+#if PG_VERSION_NUM >= PG_VERSION_16
+	Datum icurulesDatum = heap_getattr(tup, Anum_pg_database_daticurules, tupdesc,
+									   &isNull);
+	if (!isNull)
+	{
+		info.daticurules = TextDatumGetCString(icurulesDatum);
+	}
+#endif
+
+	table_close(rel, AccessShareLock);
+	heap_freetuple(tup);
+
+	return info;
+}
+
+
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+/*
+ * GetLocaleProviderString gets the datlocprovider stored in pg_database
+ * and returns the string representation of the datlocprovider
+ */
+static char *
+GetLocaleProviderString(char datlocprovider)
+{
+	switch (datlocprovider)
+	{
+		case 'c':
+		{
+			return "libc";
+		}
+
+		case 'i':
+		{
+			return "icu";
+		}
+
+		default:
+		{
+			ereport(ERROR, (errmsg("unexpected datlocprovider value: %c",
+								   datlocprovider)));
+		}
+	}
+}
+
+
+#endif
+
+
+/*
+ * GenerateCreateDatabaseStatementFromPgDatabase gets the pg_database tuple and returns the
+ * CREATE DATABASE statement that can be used to create given database.
+ *
+ * Note that this doesn't deparse OID of the database and this is not a
+ * problem as we anyway don't allow specifying custom OIDs for databases
+ * when creating them.
+ */
+static char *
+GenerateCreateDatabaseStatementFromPgDatabase(Form_pg_database databaseForm)
```

Comment:
This function doesn't put any efforts to propagate what `datacl` field says. Probably this cannot be done by CREATE DATABASE but we probably need to send some other set of DDL commands to set ACL for the databases properly, need to understand this.

Reply:
Fixed could you check the code and tests?